scripts/pathseq/wdl/pathseq_pipeline.wdl

```diff
@@ -20,179 +26,712 @@
+    Float? frac_non_host_unpaired = ProcessFilterMetrics.frac_final_unpaired
+    Float? frac_non_host_total =ProcessFilterMetrics.frac_final_total
+    Float? frac_qual_cpx_filtered = ProcessFilterMetrics.frac_qual_cpx_filtered
+    Float? frac_host_filtered = ProcessFilterMetrics.frac_host_filtered
+    Float? frac_dup_filtered = ProcessFilterMetrics.frac_dup_filtered
+
+    # Only generated if downsample=true
+    Int? total_reads = Downsample.total_reads
+  }
+}
+
+task CramToBam {
+  File cram_file
+  File? cram_index
+  File reference_fasta
+  File? reference_index
+
+  String docker
+
+  Int? cpu = 4
+  Float? mem_gb = 15
+  Int? extra_disk_gb = 10
+  Int? preemptible = 3
+  Int? max_retries = 1
+
+  String bam_file_name = basename(cram_file, ".cram") + ".bam"
+
+  File cram_index_file = select_first([cram_index, cram_file + ".crai"])
+  File reference_index_file = select_first([reference_index, reference_fasta + ".fai"])
+
+  Float cram_inflate_ratio = 3.0
+  Float cram_size = size(cram_file, "GiB")
+  Float cram_index_size = size(cram_index_file, "GiB")
+  Float bam_size = cram_inflate_ratio * cram_size
+  Float bam_index_size = cram_index_size
+  Float ref_size = size(reference_fasta, "GiB")
+  Float ref_index_size = size(reference_index_file, "GiB")
+  Int vm_disk_size = ceil(cram_size + cram_index_size + bam_size + bam_index_size + ref_size + ref_index_size + extra_disk_gb)
+
+  output {
+    File bam_file = bam_file_name
+    File bam_index = bam_file_name + ".bai"
+  }
+  command <<<
+
+        set -Eeuo pipefail
+
+        # covert cram to bam
+        samtools view  -@ ${cpu} -h -b -T "${reference_fasta}" -o "${bam_file_name}" "${cram_file}"
+
+        # index bam file
+        samtools index -@ ${cpu} "${bam_file_name}"
+
+  >>>
+  runtime {
+    cpu: 1
+    memory: mem_gb + " GiB"
+    disks: "local-disk " + vm_disk_size + " HDD"
+    bootDiskSizeGb: 10
+    docker: docker
+    preemptible: preemptible
+    maxRetries: max_retries
+  }
+}
+
+# Downsamples BAM to a specified number of reads
+task Downsample {
+  File input_bam_file
+  File? input_bam_index_file
+  String downsampled_bam_filename
+
+  Int reads_after_downsampling
+
+  String docker
+  Int additional_disk_gb = 20
+  Int preemptible_tries = 3
+
+  Int disk_size = ceil(size(input_bam_file, "GB")*2 + additional_disk_gb)
+
+  command <<<
+    set -euo pipefail
+    if ${defined(input_bam_index_file)}; then
+      NUM_READS=`samtools idxstats ${input_bam_file} | awk '{s+=$3+$4} END {print s}'`
+    else
+      NUM_READS=`samtools view -c ${input_bam_file}`
+    fi
+    P_DOWNSAMPLE=`python -c "print ${reads_after_downsampling}/float($NUM_READS)"`
+    RESULT=`python -c "print $P_DOWNSAMPLE > 1"`
+    if [ "$RESULT" == "True" ]
+    then
+        P_DOWNSAMPLE="1"
+    fi
+    echo $NUM_READS > num_reads.txt
+    java -Xmx2000m -jar /usr/gitc/picard.jar \
+      DownsampleSam \
+      INPUT=${input_bam_file} \
```

Comment:
To call this from GATK I think the args will go to `-I`, `-O`, `--VALIDATON_STRINGENCY` and maybe `--P`?

Reply:
Done

Reply:
Changed this to use GATK instead of Picard